cluster-autoscaler/core/scale_up.go

```diff
@@ -77,8 +78,26 @@ func ScaleUp(context *AutoscalingContext, unschedulablePods []*apiv1.Pod, nodes
 	podsRemainUnschedulable := make(map[*apiv1.Pod]bool)
 	expansionOptions := make([]expander.Option, 0)
 
-	for _, nodeGroup := range context.CloudProvider.NodeGroups() {
+	nodeGroups := context.CloudProvider.NodeGroups()
 
+	if context.AutoscalingOptions.NodeAutoprovisioningEnabled {
+		machines, err := context.CloudProvider.GetAvilableMachineTypes()
+		if err != nil {
+			glog.Warningf("Failed to get machine types: %v", err)
+		} else {
+			bestLabels := labels.BestLabelSet(unschedulablePods)
+			for _, machineType := range machines {
+				nodeGroup, err := context.CloudProvider.NewNodeGroup(machineType, bestLabels, nil)
+				if err != nil {
+					glog.Warningf("Unable to build temporary node group for %s: %v", machineType, err)
+				} else {
+					nodeGroups = append(nodeGroups, nodeGroup)
```

Comment:
Maybe we should filter out autoprovisioned nodeGroups identical to existing node groups? Granted, they will lose in pricing expander later on, but it will be pretty embarrassing if someone ever uses a different expander.

Reply:
It is not obvious that we should filter them. If an existing just-autoscaled group has a limit and the best option is to create yet another one then we should do this. Expander authors should handle non-existing node groups appropriately. ScaleUp is, as you wrote, complex enough.